csharp/extractor/Semmle.Extraction.CSharp.Standalone/Razor.cs

```diff
@@ -0,0 +1,99 @@
+    {
+        private readonly DotnetVersion sdk;
+        private readonly ProgressMonitor progressMonitor;
+        private readonly DotNet dotNet;
+        private readonly string sourceGeneratorFolder;
+        private readonly string cscPath;
+
+        public Razor(DotnetVersion sdk, DotNet dotNet, ProgressMonitor progressMonitor)
+        {
+            this.sdk = sdk;
+            this.progressMonitor = progressMonitor;
+            this.dotNet = dotNet;
+
+            sourceGeneratorFolder = Path.Combine(this.sdk.FullPath, "Sdks/Microsoft.NET.Sdk.Razor/source-generators");
+            if (!Directory.Exists(sourceGeneratorFolder))
+            {
+                this.progressMonitor.RazorSourceGeneratorMissing(sourceGeneratorFolder);
+                throw new Exception($"Razor source generator folder {sourceGeneratorFolder} does not exist.");
+            }
+
+            cscPath = Path.Combine(this.sdk.FullPath, "Roslyn/bincore/csc.dll");
+            if (!File.Exists(cscPath))
+            {
+                this.progressMonitor.CscMissing(cscPath);
+                throw new Exception($"csc.dll {cscPath} does not exist.");
+            }
+        }
+
+        private static void GenerateAnalyzerConfig(string[] cshtmls, string analyzerConfigPath)
+        {
+            using var sw = new StreamWriter(analyzerConfigPath);
+            sw.WriteLine("is_global = true");
+
+            foreach (var f in cshtmls)
+            {
+                sw.WriteLine($"\n[{f}]");
+                var base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(f)); // TODO: this should be the relative path of the file.
+                sw.WriteLine($"build_metadata.AdditionalFiles.TargetPath = {base64}");
+            }
+        }
+
+        public IEnumerable<string> GenerateFiles(string[] cshtmls, string workingDirectory)
+        {
+            // TODO: the below command might be too long. It should be written to a temp file and passed to csc via @.
+
+            var name = Guid.NewGuid().ToString("N").ToUpper();
+            var analyzerConfig = Path.Combine(Path.GetTempPath(), name + ".txt");
+            var dllPath = Path.Combine(Path.GetTempPath(), name + ".dll");
+            var outputFolder = Path.Combine(workingDirectory, name);
+            Directory.CreateDirectory(outputFolder);
+
+            try
+            {
+                GenerateAnalyzerConfig(cshtmls, Path.Combine(sourceGeneratorFolder, analyzerConfig));
+
+                var args = new StringBuilder();
+                args.Append($"{cscPath} --do-not-trace /target:exe /generatedfilesout:{outputFolder} /out:{dllPath} /analyzerconfig:{analyzerConfig} ");
+                foreach (var f in Directory.GetFiles(sourceGeneratorFolder, "*.dll"))
+                {
+                    args.Append($"/analyzer:{f} ");
+                }
+                foreach (var f in cshtmls)
+                {
+                    args.Append($"/additionalfile:{f} ");
+                }
```

Comment:
We need to pass at least some aspnet dlls as references. I've noticed that the rendered output through this call, and through a traced build differs, because standard tag helpers are not used in the standalone extraction. We at least need `Microsoft.AspNetCore.Mvc.TagHelpers.dll`, but possibly we could include all references that were anyways collected for the compilation.